### ISSUE TITLE
a `con` to list for tree-based models (or room for improvement??)

### ISSUE THREAD

**Issue by frankiethull** [tidymodels/orbital#97]
I think one thing that would be useful to include in [Pros & Cons](https://orbital.tidymodels.org/articles/pros-cons.html#cant-guarantee-predictions-speed) is specifying that prediction speeds begin to slow down above 100 trees in a boosted tree model. We have many models with 300+ trees. I think the real issue with speed is the (very large) SQL statement formula. Not sure if this is something that can be sped up since it's going to be a large formula. But wanted to give my user feedback after the demo today! 

(or let me know if this is more of a *me* issue!)

Thanks again to the `orbital` team! 

``` r
set.seed(314159)
################################################################
#                       xgboost orb
################################################################

# library(bonsai) # no lightgbm support
library(parsnip) # xgboost
library(orbital)
library(duckdb)
#> Loading required package: DBI

model_df <- modeldata::penguins |> na.omit()

# large amount of trees (above 100 trees)
xgb_spec <- parsnip::boost_tree(trees = 200) |>
        parsnip::set_engine("xgboost") |>
        parsnip::set_mode("regression")

xgb_rec <- recipes::recipe(body_mass_g ~ ., model_df) |>
           recipes::step_dummy(recipes::all_nominal_predictors()) |>
           recipes::step_nzv(recipes::all_predictors())


xgb_wkflw <- workflows::workflow() |>
             workflows::add_model(xgb_spec) |>
             workflows::add_recipe(xgb_rec)

xgb_fit <- xgb_wkflw |>
          workflows::fit(model_df)

# conversion begins to slow down above ~100 trees
xgb_orb <- orbital::orbital(xgb_fit)

################################################################
#                    in memory duckdb
################################################################

penquack_con <- dbConnect(duckdb(dbdir = ":memory:"))
model_table <- dplyr::copy_to(penquack_con, model_df)

#  complex formula is a bit slower which slows down the predictions
# xgb_sql <- orbital::orbital_sql(xgb_orb, penquack_con)

# showcase predict speed with many trees

### w/ orbital
tictoc::tic()
predict(xgb_orb, model_table)
#> # Source:   SQL [?? x 1]
#> # Database: DuckDB v1.3.2 [frankiethull@Linux 6.12.10-76061203-generic:R 4.5.1/:memory:]
#>    .pred
#>    <dbl>
#>  1 3750.
#>  2 3799.
#>  3 3250.
#>  4 3450.
#>  5 3650.
#>  6 3625.
#>  7 4675.
#>  8 3200.
#>  9 3800.
#> 10 4400.
#> # ℹ more rows
tictoc::toc()
#> 82.894 sec elapsed

### w/o orbital
tictoc::tic()
predict(xgb_fit, model_df)
#> # A tibble: 333 × 1
#>    .pred
#>    <dbl>
#>  1 3750.
#>  2 3799.
#>  3 3250.
#>  4 3450.
#>  5 3650.
#>  6 3625.
#>  7 4675.
#>  8 3200.
#>  9 3800.
#> 10 4400.
#> # ℹ 323 more rows
tictoc::toc()
#> 0.013 sec elapsed
```

<sup>Created on 2025-07-30 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by EmilHvitfeldt** [comment]
Hello @frankiethull 👋 

Good news, bad news!

Bad news is that the orbital translation can't be improved further, as it has to faithfully replicate the original model. Also for data this size it will be hard to beat the "in R" version.

But we have to consider that the orbital version doesn't need tidymodels or xgboost loaded at prediction time. You just need orbital (which is relatively light) and duckdb. That docker container is going to be quite a bit smaller. you can get even smaller if you save the expression from orbital to a file and not even load orbital.

You can likely get some benefits by using duckdb as duckdb allows for "not in memory data sets" which we can predict on as well. Speed will not always win, but when we have seen speed improvements it has typically been when data was getting so big that the transferring started becoming significant. 

Good news is that we can do a couple of things to improve the runtime. Mainly that would involve less complex models. As a quick example below I set `tree_depth = 2` down from `6` which is the default. That game me a orbital fit increase down to >4 seconds, despite the formula only decreasing by 30%.

<details>
<summary>`tree_depth = 2` reprex</summary>

``` r
set.seed(314159)
################################################################
#                       xgboost orb
################################################################

# library(bonsai) # no lightgbm support
library(parsnip) # xgboost
library(orbital)
library(duckdb)
#> Loading required package: DBI
#> Loading required package: DBI

model_df <- modeldata::penguins |> na.omit()

# large amount of trees (above 100 trees)
xgb_spec <- parsnip::boost_tree(trees = 200, tree_depth = 2) |>
        parsnip::set_engine("xgboost") |>
        parsnip::set_mode("regression")

xgb_rec <- recipes::recipe(body_mass_g ~ ., model_df) |>
           recipes::step_dummy(recipes::all_nominal_predictors()) |>
           recipes::step_nzv(recipes::all_predictors())


xgb_wkflw <- workflows::workflow() |>
             workflows::add_model(xgb_spec) |>
             workflows::add_recipe(xgb_rec)

xgb_fit <- xgb_wkflw |>
          workflows::fit(model_df)

# conversion begins to slow down above ~100 trees
xgb_orb <- orbital::orbital(xgb_fit)

################################################################
#                    in memory duckdb
################################################################

penquack_con <- dbConnect(duckdb(dbdir = ":memory:"))
model_table <- dplyr::copy_to(penquack_con, model_df)

stringr::str_split(xgb_orb[6], "dplyr::case_when")[[1]] |> 
  purrr::map_int(nchar) |>
  summary()
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>     4.0   379.0   393.0   380.5   401.0   408.0

#  complex formula is a bit slower which slows down the predictions
# xgb_sql <- orbital::orbital_sql(xgb_orb, penquack_con)

# showcase predict speed with many trees

### w/ orbital
tictoc::tic()
predict(xgb_orb, model_table)
#> # Source:   SQL [?? x 1]
#> # Database: DuckDB v1.2.1 [root@Darwin 24.5.0:R 4.5.0/:memory:]
#>    .pred
#>    <dbl>
#>  1 3729.
#>  2 3631.
#>  3 3490.
#>  4 3453.
#>  5 3818.
#>  6 3476.
#>  7 4340.
#>  8 3228.
#>  9 3825.
#> 10 4299.
#> # ℹ more rows
tictoc::toc()
#> 3.536 sec elapsed

### w/o orbital
tictoc::tic()
predict(xgb_fit, model_df)
#> # A tibble: 333 × 1
#>    .pred
#>    <dbl>
#>  1 3729.
#>  2 3631.
#>  3 3490.
#>  4 3453.
#>  5 3818.
#>  6 3476.
#>  7 4340.
#>  8 3228.
#>  9 3825.
#> 10 4299.
#> # ℹ 323 more rows
tictoc::toc()
#> 0.008 sec elapsed
```

<sup>Created on 2025-07-30 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

</details>

While i was writing this response I had an idea. do you think a vignette along the lines off "how to fit models to optimize for orbital speed"? 
Xgboost has a lot of plateaus, and `trees` and `tree_depth` are the main arguments that affect orbital size, so by setting `learning_rate` and other things you can get the first ones quite low with minimal performance loss.

**Comment by frankiethull** [comment]
I appreciate your response! I will have to test a few ways to lower complexity, we are cv-hyperparameter finetuning so it can be a complex model. I will also need to try the non-memory duckdb tests. We are typically running [8760 (hourly rows) * 100 (iters) * 25 (years)] sims per model so I like to hear the speed up on larger datasets! 

I also like your idea of a new vignette/article. I felt icky writing this as a "con" anyway. I'd like to spin it off as a learning/teaching opportunity for users. Appreciate all your help on this!

**Comment by emilyriederer** [comment]
@frankiethull -- I have mostly been using the python version of `orbital` so I can't speak to exactly how the code is rendered in R. However, depending on how good your prod database is at optimizing queries, I think there *are* ways you can improve runtime for tree models! Ensemble models are "embarrassingly parallel" but sometimes the SQL code un-parallelizes. 

Three things I've noticed from the python version. I think the R version may not do (1) and (2), but I suspect (3) might help:

1. **Feature engineering code is repeated every time that feature is used (versus preprocessed in CTE).** I don't this the R version does this based on the vignettes, but if so, I favor generating/deploying separate scripts for features + predictions anyway. 

2. **All logic is repeated for positive case probability, negative case probability, and class probability.** At least in `duckdb`, the optimizer doesn't recognize that the core logic is repeated. Again, I don't think the R version does this, but I've been stripping out only the positive-class probability which improves runtime. 

3. **Trees code is all included in one big SQL blob leading up to the prediction which kills the potentially parallel calculations for individual trees.** I can't tell if the R version does this, so maybe this one would help you? I (very non-rigorously) tested moving the individual trees to be calculated as separate columns in a CTE. This lets the database know they can happen in parallel and makes the final calculation much smaller. Example is [here](https://github.com/emilyriederer/orbital-exploration/blob/main/orbital-big.ipynb) which in a (extremely non-rigorous, cannot emphasize this enough) test, roughly halved scoring time. Query post-processing is in chunk 6 and timing is in chunks 10-11. This is for a Random Forest, so obviously you have to be slightly more cautious editing the query for a GBM, but conceptually there's no reason it shouldn't work. 

I think optimizations are probably model-specific and database-specific, so it's possible the `orbital` team can't take these on. However, just wanted to share some thoughts if you want to do your own post-processing. Again, in python, I'm using the `sqlglot` package to make these changes to the SQL code's AST; not sure if there is a similar R tool.

**Comment by EmilHvitfeldt** [comment]
1. Feature engineering code is repeated every time that feature is used (versus preprocessed in CTE)

This should not happen, and if so should be reported as a bug. I'm aware that some of the tools used in the python version does this automatically and we are working on a way to avoid it.

2. All logic is repeated for positive case probability, negative case probability, and class probability.

That is correct in the binary case here for xgboost at least.

``` r
library(tidymodels)
library(orbital)

iris$Species <- factor(iris$Species == "setosa")

mod_spec <- boost_tree(trees = 1, tree_depth = 1) |>
  set_mode("classification") |>
  set_engine("xgboost")

wf_spec <- workflow() |>
  add_formula(Species ~ .) |>
  add_model(mod_spec)

wf_fit <- fit(wf_spec, iris)

orb <- orbital(wf_fit, type = "prob")

print(orb, truncate = FALSE)
#> 
#> ── orbital Object ──────────────────────────────────────────────────────────────
#> • .pred_FALSE = 1 - 1/(1 + exp(0 + dplyr::case_when((Petal.Length < 2.45 |
#> is.na(Petal.Length)) ~ -0.5555556, Petal.Length >= 2.45 ~ 0.5769231) +
#> log(0.5/(1 - 0.5))))
#> • .pred_TRUE = 1 - (1 - 1/(1 + exp(0 + dplyr::case_when((Petal.Length < 2.45 |
#> is.na(Petal.Length)) ~ -0.5555556, Petal.Length >= 2.45 ~ 0.5769231) +
#> log(0.5/(1 - 0.5)))))
#> ────────────────────────────────────────────────────────────────────────────────
#> 2 equations in total.
```

<sup>Created on 2025-08-17 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

however the multi-class setting should be fairly optimized 

``` r
library(tidymodels)
library(orbital)

mod_spec <- boost_tree(trees = 1, tree_depth = 1) |>
  set_mode("classification") |>
  set_engine("xgboost")

wf_spec <- workflow() |>
  add_formula(Species ~ .) |>
  add_model(mod_spec)

wf_fit <- fit(wf_spec, iris)

orb <- orbital(wf_fit, type = "prob")

print(orb, truncate = FALSE)
#> 
#> ── orbital Object ──────────────────────────────────────────────────────────────
#> • setosa = 0 + dplyr::case_when((Petal.Length < 2.45 | is.na(Petal.Length)) ~
#> 0.430622, Petal.Length >= 2.45 ~ -0.2200489)
#> • versicolor = 0 + dplyr::case_when((Petal.Length < 2.45 | is.na(Petal.Length))
#> ~ -0.215311, Petal.Length >= 2.45 ~ 0.1100244)
#> • virginica = 0 + dplyr::case_when((Petal.Width < 1.65 | is.na(Petal.Width)) ~
#> -0.1942446, Petal.Width >= 1.65 ~ 0.4029851)
#> • norm = exp(setosa) + exp(versicolor) + exp(virginica)
#> • .pred_setosa = exp(setosa) / norm
#> • .pred_versicolor = exp(versicolor) / norm
#> • .pred_virginica = exp(virginica) / norm
#> ────────────────────────────────────────────────────────────────────────────────
#> 7 equations in total.
```

<sup>Created on 2025-08-17 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

3. Trees code is all included in one big blob of SQL logic defining the prediction column which may limit some parallel execution of the individual trees.

So @emilyriederer, you are saying that we would be able to gain some speed by splitting the trees up and adding them together at the end?


Do this
```r
tree_1 <- dplyr::case_when(
  (Petal.Length < 2.45 |
    is.na(Petal.Length)) ~
    -0.5555556,
  Petal.Length >= 2.45 ~ 0.5769231
)

tree_2 <- dplyr::case_when(
    (Petal.Length < 2.45 | is.na(Petal.Length)) ~ -0.4346052,
    Petal.Length >= 2.45 ~ 0.4489912
  )

tree_3 <- dplyr::case_when(
    (Petal.Length < 2.45 |
      is.na(Petal.Length)) ~
      -0.3736209,
    Petal.Length >= 2.45 ~ 0.3875872
  )

all_trees <- tree_1 + tree_2 + tree_3
```

Instead of 

```r
all_trees <- dplyr::case_when(
  (Petal.Length < 2.45 |
    is.na(Petal.Length)) ~
    -0.5555556,
  Petal.Length >= 2.45 ~ 0.5769231
) +
  dplyr::case_when(
    (Petal.Length < 2.45 | is.na(Petal.Length)) ~ -0.4346052,
    Petal.Length >= 2.45 ~ 0.4489912
  ) +
  dplyr::case_when(
    (Petal.Length < 2.45 |
      is.na(Petal.Length)) ~
      -0.3736209,
    Petal.Length >= 2.45 ~ 0.3875872
  )
```

If so, that should be possible for us to do
